File: golang/state/fruit.go
package state

type fruit struct {
	stage Stage
}

func NewFruit() fruit {
	//幼苗状态开始
	seedling := seedling{}
	fruit := fruit{seedling}
	return fruit
}

func (f *fruit) Water() {
	f.stage.water()
	if _, ok := (f.stage).(seedling); ok {
		bloom := bloom{}
		f.stage = bloom
	} else {
		maturity := maturity{}
		f.stage = maturity
	}
}

func (f *fruit) Harvest() {
	f.stage.harvest()
	if _, ok := f.stage.(maturity); ok {
		seedling := seedling{}
		f.stage = seedling
	}
}

File: java/src/state/Fruit.java
package state;


public class Fruit {
    private  IState state;

    public Fruit(IState state){
        this.state = state;
    }

    public void water(){
        this.state.water();
        if (state instanceof Seedling){
           Bloom bloom = new Bloom();
           this.state = bloom;
        } else{
            maturity maturity = new maturity();
            this.state = maturity;
        }
    }

    public void harvest(){
        this.state.harvest();
        if (state instanceof maturity){
            Seedling seedling = new Seedling();
            this.state = seedling;
        }
    }

}
